runtime/interpreter/storage.go

```diff
@@ -33,31 +33,42 @@ import (
 	"github.com/onflow/cadence/runtime/common"
 )
 
-func StoredValue(storable atree.Storable, storage atree.SlabStorage) Value {
+func StoredValue(gauge common.MemoryGauge, storable atree.Storable, storage atree.SlabStorage) Value {
 	storedValue, err := storable.StoredValue(storage)
 	if err != nil {
 		panic(err)
 	}
 
-	return MustConvertStoredValue(storedValue)
+	return MustConvertStoredValue(gauge, storedValue)
 }
 
-func MustConvertStoredValue(value atree.Value) Value {
-	converted, err := ConvertStoredValue(value)
+func MustConvertStoredValue(gauge common.MemoryGauge, value atree.Value) Value {
+	converted, err := ConvertStoredValue(gauge, value)
 	if err != nil {
 		panic(err)
 	}
 	return converted
 }
 
-func ConvertStoredValue(value atree.Value) (Value, error) {
+func MustConvertUnmeteredStoredValue(value atree.Value) Value {
+	converted, err := ConvertStoredValue(nil, value)
+	if err != nil {
+		panic(err)
+	}
+	return converted
+}
+
+func ConvertStoredValue(gauge common.MemoryGauge, value atree.Value) (Value, error) {
 	// TODO: Meter container values created below.
 	//   This is currently technically challenging because, this method is used in various
 	//   places directly/indirectly, where the interpreter instance is not available.
 	//   e.g: Iterator(), value walk(), RecursiveString(), String() method of values, etc.
 
 	switch value := value.(type) {
 	case *atree.Array:
+		if gauge != nil {
+			gauge.UseMemory(common.NewConstantMemoryUsage(common.MemoryKindArray))
+		}
 		return &ArrayValue{
 			Type:  value.Type().(ArrayStaticType),
 			array: value,
```

Comment:
Maybe we should have constructors (private?) for the container values that are constructed from their underlying atree values.

For example here, it might make sense to add a constructor function `func newArrayValueFromAtreeValue(array *atree.Array, memoryGauge common.MemoryGauge)` and call it here.

`NewArrayValueWithIterator` would also call this function at the end of its body. The metering in `NewArrayValueWithIterator` would also move to the new `newArrayValueFromAtreeValue`.

That way we only meter in one place, the constructor